src/senseflow/binary.rs

```diff
@@ -0,0 +1,229 @@
+use std::convert::TryInto;
+
+pub trait FromBytes<T> {
+    fn from_le_bytes(data: &[u8]) -> T;
+    fn from_be_bytes(data: &[u8]) -> T;
+}
+
+macro_rules! impl_from_bytes {
+    ($($t:ty),+) => {
+        $(impl FromBytes<$t> for $t {
+            fn from_le_bytes(data: &[u8]) -> $t {
+                <$t>::from_le_bytes(data.try_into().unwrap())
+            }
+            fn from_be_bytes(data: &[u8]) -> $t {
+                <$t>::from_be_bytes(data.try_into().unwrap())
+            }
+        })+
+    }
+}
+
+impl_from_bytes!(u8, u16, i16, i32, u32, i64,f32,f64);
+fn from_bytes<T: FromBytes<T>>(data: &[u8]) -> T {
+    T::from_le_bytes(&data[..std::mem::size_of::<T>()])
+}
+
+
+pub fn parse<T: Read + Seek>(stream: &mut T, _size: usize) -> Result<Vec<SampleInfo>> {
+    let mut stream = std::io::BufReader::new(stream);
+
+    let mut gyro = Vec::new();
+    let mut accl = Vec::new();
+    let mut angl = Vec::new();
+    let mut magn = Vec::new();
+    let mut quat = Vec::new();
+
+    let mut last_timestamp;
+
+    let mut buf = vec![0u8; 512];
+    stream.read_exact(&mut buf[0..512])?;
+
+    let brand =  unsafe { std::str::from_utf8_unchecked(&buf[0..12])};
+    let version =  unsafe { std::str::from_utf8_unchecked(&buf[12..16])};
+    let _product_id = &buf[16..36];
+    let _product_sn = &buf[36..52];
+
+    let imu_orientation =   std::str::from_utf8(&buf[60..64]).unwrap_or("XYZ");
+    let yy = buf[64] as i32 + 2000;
+    let mm = buf[65] as u32;
+    let dd = buf[66] as u32;
+    let h  = buf[67] as u32;
+    let m  = buf[68] as u32;
+    let s  = buf[69] as u32;
+    let ms = 0u32;
+    let create_at =  chrono::NaiveDate::from_ymd_opt(yy, mm, dd).and_then(|x| x.and_hms_milli_opt(h, m, s, ms)).unwrap_or_default();
+    let first_timestamp = 0f64;//create_at.timestamp_millis() as f64 / 1000.0;
+
+    let _init_quad = TimeQuaternion {
+        t: (first_timestamp*1000.0) as f64,
+        v: Quaternion{
+            w: from_bytes::<f32>(&buf[76..80]) as f64,
+            x: from_bytes::<f32>(&buf[80..84]) as f64,
+            y: from_bytes::<f32>(&buf[84..88]) as f64,
+            z: from_bytes::<f32>(&buf[88..92]) as f64,
+        }
+    };
+
+    let log_freq = from_bytes::<u32>(&buf[92..96]);
+
+    let acc_odr = from_bytes::<u16>(&buf[144..146]);
+    let acc_max_bw = from_bytes::<u16>(&buf[146..148]);
+    let acc_timeoffset = from_bytes::<i32>(&buf[148..152]);
+    let acc_range = from_bytes::<u32>(&buf[152..156]) as f64;
+
+    let gyro_odr = from_bytes::<u16>(&buf[156..158]);
+    let gyro_max_bw = from_bytes::<u16>(&buf[158..160]);
+    let gyro_timeoffset = from_bytes::<i32>(&buf[160..164]);
+    let gyro_range = from_bytes::<u32>(&buf[164..168])as f64;
+
+    let mag_odr = from_bytes::<u16>(&buf[168..170]);
+    let mag_max_bw = from_bytes::<u16>(&buf[170..172]);
+    let mag_timeoffset = from_bytes::<i32>(&buf[172..176]);
+    let mag_range = from_bytes::<u32>(&buf[176..180])as f64 / 1000.0;
```

Comment:
Instead of all this `from_bytes` with exact offsets, use `let mut d = std::io::Cursor::new(buf);` and then `d.read_u16::<LittleEndian>()?` functions

Reply:
Some of the bytes in a sector(512Bytes) are reserved, and if use the "read" method, i need to manually read the reserved bytes but not use them.Is there a way to cross these reserved bytes?

Reply:
sure, just use `d.seek(SeekFrom::Current(8))?;` for example to skip 8 bytes

Reply:
Get it. It's the better function to format bytes!